src/estimagic/optimization/subsolvers/_trsbox_quadratic.py

```diff
@@ -0,0 +1,624 @@
+            tangent, sine, s_hess_s, x_hess_x, x_hess_s, x_grad, s_norm
+        )
+
+        if new_reduction > max_reduction:
+            max_reduction = new_reduction
+            index_angle_greatest_reduction = i
+            previous_reduction = old_reduction
+        elif i == index_angle_greatest_reduction + 1:
+            next_reduction = new_reduction
+        old_reduction = new_reduction
+
+    return (
+        previous_reduction,
+        next_reduction,
+        max_reduction,
+        tangent,
+        index_angle_greatest_reduction,
+        n_angles,
+    )
+
+
+def _update_candidate_vectors_and_reduction_alt_step(
+    x_candidate,
+    search_direction,
+    x_bounded,
+    gradient_candidate,
+    cosine,
+    sine,
+    hess_s,
+    hessian_reduced,
+):
+    """Update candidate vectors and the associated criterion reduction.
+
+    If the angle of the alternative iteration is restricted by a bound on a
+    free variable, that variable is fixed at the bound.
+    """
+    gradient_candidate += (cosine - 1.0) * hessian_reduced + sine * hess_s
+    x_candidate[x_bounded == 0] = (
+        cosine * x_candidate[x_bounded == 0] + sine * search_direction[x_bounded == 0]
+    )
+    x_grad = x_candidate[x_bounded == 0] @ gradient_candidate[x_bounded == 0]
+    gradient_reduced = _calc_sum_sq(gradient_candidate[x_bounded == 0])
+    hessian_reduced = cosine * hessian_reduced + sine * hess_s
+
+    return x_candidate, gradient_candidate, x_grad, gradient_reduced, hessian_reduced
+
+
+def _compute_new_search_direction_and_norm(
+    x_candidate, x_bounded, x_reduced, gradient_candidate, x_grad, raw_reduction
+):
+    """Compute the new search direction and its norm."""
+    raw_reduction = np.sqrt(raw_reduction)
+    search_direction = np.zeros_like(x_candidate)
+
+    search_direction[x_bounded == 0] = (
+        x_grad * x_candidate[x_bounded == 0]
+        - x_reduced * gradient_candidate[x_bounded == 0]
+    ) / raw_reduction
+    s_norm = -raw_reduction
+
+    return search_direction, s_norm
+
+
+def _calc_new_reduction(tangent, sine, s_hess_s, x_hess_x, x_hess_s, x_grad, s_norm):
+    """Calculate the new reduction in the criterion function."""
+    raw_reduction = s_hess_s + tangent * (tangent * x_hess_x - 2.0 * x_hess_s)
+    current_reduction = sine * (tangent * x_grad - s_norm - 0.5 * sine * raw_reduction)
+
+    return current_reduction
+
+
+def _update_tangent(
+    index_angle_greatest_reduction,
+    bound_on_tangent,
+    n_angles,
+    next_reduction,
+    previous_reduction,
+    max_reduction,
+):
+    """Update the tangent of half the angle to the trust-region boundary."""
+    raw_reduction = (next_reduction - previous_reduction) / (
+        2.0 * max_reduction - previous_reduction - next_reduction
+    )
+    tangent = (
+        bound_on_tangent
+        * ((index_angle_greatest_reduction + 1) + 0.5 * raw_reduction)
+        / n_angles
+    )
+    return tangent
+
+
+def _calc_sum_sq(x):
```

Comment:
You don't need the transpose here and it's questionable if this needs to be a function at al. In most cases just `x @ x` is shorter and clearer than calling this function.